File: vane-core/src/main/java/org/oddlama/vane/util/WorldUtil.java
package org.oddlama.vane.util;

import java.util.HashMap;
import java.util.UUID;
import org.bukkit.World;
import org.bukkit.plugin.Plugin;
import org.bukkit.scheduler.BukkitTask;

public class WorldUtil {

    private static final HashMap<UUID, BukkitTask> runningTimeChangeTasks = new HashMap<>();

    public static boolean changeTimeSmoothly(
        final World world,
        final Plugin plugin,
        final long worldTicks,
        final long interpolationTicks
    ) {
        synchronized (runningTimeChangeTasks) {
            if (runningTimeChangeTasks.containsKey(world.getUID())) {
                return false;
            }

            // Calculate relative time from and to
            var relTo = worldTicks;
            var relFrom = world.getTime();
            if (relTo <= relFrom) {
                relTo += 24000;
            }

            // Calculate absolute values
            final var deltaTicks = relTo - relFrom;
            final var absoluteFrom = world.getFullTime();
            final var absoluteTo = absoluteFrom - relFrom + relTo;

            // Task to advance time every tick
            BukkitTask task = plugin
                .getServer()
                .getScheduler()
                .runTaskTimer(
                    plugin,
                    new Runnable() {
                        private long elapsed = 0;

                        @Override
                        public void run() {
                            // Remove a task if we finished interpolation
                            if (elapsed > interpolationTicks) {
                                synchronized (runningTimeChangeTasks) {
                                    runningTimeChangeTasks.remove(world.getUID()).cancel();
                                }
                            }

                            // Make the transition smooth by applying a cosine
                            var linDelta = (float) elapsed / interpolationTicks;
                            var delta = (1f - (float) Math.cos(Math.PI * linDelta)) / 2f;

                            var curTicks = absoluteFrom + (long) (deltaTicks * delta);
                            world.setFullTime(curTicks);
                            ++elapsed;
                        }
                    },
                    1,
                    1
                );

            runningTimeChangeTasks.put(world.getUID(), task);
        }

        return true;
    }
}

File: vane-core/src/main/kotlin/org/oddlama/vane/util/WorldUtil.kt
package org.oddlama.vane.util

import org.bukkit.World
import org.bukkit.plugin.Plugin
import org.bukkit.scheduler.BukkitTask
import java.util.*
import kotlin.math.PI
import kotlin.math.cos

object WorldUtil {
    private val runningTimeChangeTasks = mutableMapOf<UUID, BukkitTask>()

    @JvmStatic
    fun changeTimeSmoothly(
        world: World,
        plugin: Plugin,
        worldTicks: Long,
        interpolationTicks: Long
    ): Boolean {
        synchronized(runningTimeChangeTasks) {
            if (world.uid in runningTimeChangeTasks) return false

            val relFrom = world.time
            val relTo = if (worldTicks > relFrom) worldTicks else worldTicks + 24000
            val deltaTicks = relTo - relFrom
            val absoluteFrom = world.fullTime

            var elapsed = 0L
            val task = plugin.server.scheduler.runTaskTimer(plugin, Runnable {
                if (elapsed > interpolationTicks) {
                    synchronized(runningTimeChangeTasks) {
                        runningTimeChangeTasks.remove(world.uid)!!.cancel()
                    }
                }
                val linDelta = elapsed.toFloat() / interpolationTicks
                val delta = (1f - cos(PI * linDelta).toFloat()) / 2f
                world.fullTime = absoluteFrom + (deltaTicks * delta).toLong()
                elapsed++
            }, 1L, 1L)

            runningTimeChangeTasks[world.uid] = task
        }
        return true
    }
}
